src/main/java/com/thomashan/coup/turn/state/WaitingChallengeActionState.java

```diff
@@ -0,0 +1,154 @@
+import com.thomashan.coup.player.Player;
+import com.thomashan.coup.player.Players;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.thomashan.coup.action.ChallengeActionType.ALLOW;
+import static com.thomashan.coup.action.ChallengeActionType.CHALLENGE;
+import static java.util.Optional.empty;
+
+abstract class WaitingChallengeActionState implements TurnState<ChallengeAction> {
+    private final Players players;
+    private final Player player;
+    private final Deck deck;
+    private final MainAction mainAction;
+    private final ImmutableList<Action> actionHistory;
+    private final Optional<Player> target;
+    private final ImmutableList<Player> challengeablePlayers;
+    private final FromState fromState;
+
+    protected WaitingChallengeActionState(Players players, Player player, Deck deck, MainAction mainAction, List<Action> actionHistory, Player target, List<Player> challengeablePlayers, FromState fromState) {
+        this.players = players;
+        this.player = player;
+        this.deck = deck;
+        this.challengeablePlayers = ImmutableList.of(challengeablePlayers);
+        this.mainAction = mainAction;
+        this.actionHistory = ImmutableList.of(actionHistory);
+        this.fromState = fromState;
+
+        if (target != null) {
+            this.target = Optional.of(target);
+        } else {
+            this.target = empty();
+        }
+    }
+
+    @Override
+    public List<Action> getActionHistory() {
+        return actionHistory;
+    }
+
+    @Override
+    public Deck getDeck() {
+        return deck;
+    }
+
+    @Override
+    public Players getPlayers() {
+        return players;
+    }
+
+    @Override
+    public Player getPlayer() {
+        return player;
+    }
+
+    @Override
+    public Optional<Player> getTarget() {
+        return target;
+    }
+
+    @Override
+    public MainAction getMainAction() {
+        return mainAction;
+    }
+
+    @Override
+    public boolean isComplete() {
+        return false;
+    }
+
+    @Override
+    public boolean isCheckAllowableActions() {
+        return true;
+    }
+
+    @Override
+    public List<ChallengeAction> getAllowableActions() {
+        return getChallengeablePlayers().stream()
+                .flatMap(p -> Stream.of(ChallengeAction.of(p, ALLOW), ChallengeAction.of(p, CHALLENGE)))
+                .collect(Collectors.toList());
+    }
+
+    protected abstract TurnState toWaitingChallengeActionState(ChallengeAction action, List<Action> actionHistory);
+
+    @Override
+    public TurnState performAction(ChallengeAction action) {
+        checkAction(action);
+
+        List<Action> newActionHistory = actionHistory.plus(action);
+
+        if (action.getChallengeActionType() == CHALLENGE) {
+            return toRevealCardState(action, newActionHistory);
+        }
+
+        if (getChallengeablePlayers().size() > 1) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid using Literals in Conditional Statements](https://app.codacy.com/app/thomashan/coup-java/pullRequest?prid=4561385)